python/private/bzlmod/pip.bzl

```diff
@@ -344,6 +362,16 @@ def _pip_impl(module_ctx):
 
 def _pip_parse_ext_attrs():
     attrs = dict({
+        "envsubst": attr.string_list(
+            mandatory = False,
+            doc = """\
+A list of environment variables to substitute (e.g. `["PIP_INDEX_URL",
+"PIP_RETRIES"]`). The corresponding variables are expanded in `extra_pip_args`
```

Comment:
@vonschultz, could you elaborate a little on what is the usecase here? Is it only for the requirements files that your project owns? What should we do for the `PIP_INDEX_URL` in a non-root module? On one hand I think that the root module should be able to override the index url + extra index urls for everyone, but if that is what we want to achieve with this PR, then maybe having support for this in:
```
pip.override(
    index_url = "foo",
    extra_index_urls = ["bar", "baz"],
)
```
might be maybe cleaner?

We could append the `index_url` and prepend the `extra_index_urls` to the extra pip args and those would take precedence (I think).

Are there other env var names that you want to expand here?

Reply:
The task I need it for is converting a project from `WORKSPACE.bazel` to `MODULE.bazel` in a case where the `WORKSPACE.bazel` uses environment variables (by writing them to a file using a repo rule and loading that file). Since loading isn't supported in `MODULE.bazel` I need the support for environment variables here instead. I have only been thinking in terms of the root module, so I'm not entirely sure what the desired behavior would be in a non-root module.

In principle, there are as many potential use cases as there are flags to `pip`, but to focus on the ones explicitly mentioned in the patch:

- `PIP_RETRIES`, and also `PIP_TIMEOUT` for that matter, are useful when the user has an unstable Internet connection. Rather like Bazel's own [--http_timeout_scaling](https://bazel.build/reference/command-line-reference#flag--http_timeout_scaling), which a user might put in their `~/.bazelrc`, it is a property of the environment we run in, rather than the repo itself.
- `PIP_INDEX_URL` is given to us by our CI system, and is pointing to an AWS CodeArtifact URL. The `PIP_INDEX_URL` contains a short-lived token for authentication, so we need to read it from the environment rather than hardcoding it in the `MODULE.bazel` file. 

I haven't been thinking of `pip.override`. Isn't that specific for a given Python package (distribution)? 
https://rules-python.readthedocs.io/en/latest/api/extensions/pip.html#pip-override
It doesn't sound like it's something you'd use to override the index url.

I'm not convinced overriding the index url for everyone (including non-root modules) is the desired behavior. If someone uses the default https://pypi.org/simple certainly, we'd like to be able to redirect that, but if a non-root module very explicitly and purposefully uses a different index url containing a small set of packages not found on the public PyPI, then it's not a given that we should override that.

(Incidentally, I think `extra_index_urls` is a huge security issue, and should never be used. That's how we got hit by https://pytorch.org/blog/compromised-nightly-dependency/, and I'm sure there are many other cases like it.)

Reply:
Apart from the environment variables I mentioned, I can also imagine `PIP_PROXY`  being useful.

Reply:
Thanks for elaborating on the extra index url usage.

I am wondering if the right thing here would be to have an `attr` for each thing `pip` supports - the downside would be that we would need to have stuff in sync with the `pip` version we support/use, but the upside would be that it would be more structured and we could more easily ensure that sensitive info does not leak into the lock file.

What do you think? How many extra attributes would you need? @rickeylev, do you have any thoughts on this?